android/core/src/main/java/com/stadiamaps/ferrostar/core/FerrostarCore.kt

```diff
@@ -118,10 +160,45 @@ class FerrostarCore(
 
         if (controller != null) {
             _state?.update { currentValue ->
-                controller.updateUserLocation(
+                val newState = controller.updateUserLocation(
                     location = location.userLocation(),
-                    state = currentValue
+                    state = currentValue.tripState
                 )
+
+                if (newState is TripState.Navigating) {
+                    if (newState.deviation is RouteDeviation.OffRoute) {
+                        if (!_routeRequestInFlight && _lastAutomaticRecalculation?.isAfter(LocalDateTime.now().minusSeconds(15)) != false) {
+                            val action = delegate?.correctiveActionForDeviation(
+                                this,
+                                newState.deviation.deviationFromRouteLine
+                            ) ?: CorrectiveAction.DoNothing()
+                            when (action) {
+                                is CorrectiveAction.DoNothing -> {
+                                    // Do nothing
+                                }
+
+                                is CorrectiveAction.GetNewRoutes -> {
+                                    _isCalculatingNewRoute = true
+                                    _scope.launch {
+                                        try {
+                                            val routes =
+                                                getRoutes(location.userLocation(), action.waypoints)
+                                            _lastAutomaticRecalculation = LocalDateTime.now()
```

Comment:
After further consideration, I've moved this down inside the finally block and done something similar in Swift to avoid the issue flagged on the call.

Reply:
After further consideration, I've moved this down inside the finally block and done something similar in Swift to avoid the issue flagged on the call.